packages/darklang/cli/cli.dark

```diff
@@ -147,5 +147,468 @@ darklang [script path]                            Run a .dark script
+    let getLocalVersion () : String =
+
+      let config =
+        (Builtin.fileRead configPath)
+        |> Stdlib.Result.map (fun config ->
+          config
+          |> Stdlib.String.fromBytesWithReplacement
+          |> Builtin.jsonParse<Dict<String>>
+          |> (Stdlib.Result.map (fun config ->
+            config
+            |> PACKAGE.Darklang.Stdlib.Dict.get "version"
+            |> Stdlib.Option.withDefault "v0.0.0")))
+
+      let localVersion =
+        match config with
+        | Ok version -> version |> Stdlib.Result.withDefault "v0.0.0"
+        | Error e -> "v0.0.0"
+
+      localVersion
+
+
+    // check if we should skip self update (if we've updated in the last 24 hours don't update)
+    let shouldSkipUpdate () : Bool =
+
+      let config =
+        (Builtin.fileRead configPath)
+        |> Stdlib.Result.map (fun config ->
+          config
+          |> Stdlib.String.fromBytesWithReplacement
+          |> Builtin.jsonParse<Dict<String>>
+          |> (Stdlib.Result.map (fun config ->
+            config
+            |> PACKAGE.Darklang.Stdlib.Dict.get "lastUpdateTimestamp"
+            |> Stdlib.Option.withDefault "0")))
+
+      let lastUpdateTimestamp =
+        match config with
+        | Ok timestamp -> timestamp |> Stdlib.Result.withDefault "0"
+        | Error e -> "0"
+
+      let now = Stdlib.DateTime.now ()
+
+      let secondsSinceLastUpdate =
+        Stdlib.DateTime.toSeconds now
+        - (lastUpdateTimestamp |> Stdlib.Int64.parse |> Builtin.unwrap)
+
+      secondsSinceLastUpdate < 86400L
+
+
+
+    // -- self update function --
+    let runSelfUpdate () : Int64 =
+      if shouldSkipUpdate () then
+        Builtin.printLine
+          "Skipping self update because we've updated in the last 24 hours"
+
+        0L
+
+      else
+        Builtin.printLine "Running self update..."
+        // get os and arch
+        let os = Stdlib.Process.runCommand "uname -s"
+        let arch = Stdlib.Process.runCommand "uname -m"
+
+        let osName = Stdlib.String.trim os.stdout
+        let archName = Stdlib.String.trim arch.stdout
+
+        // validate os and arch
+        let os =
+          if os.exitCode == 0L then
+            if osName == "Linux" then
+              "linux"
+            else if osName == "Darwin" then
+              "osx"
+            else
+              Builtin.printLine $"Unsupported OS {osName}"
+              "Unsupported"
+
+          else
+            Builtin.printLine "Failed to get OS"
+            os.exitCode
+
+        let arch =
+          if arch.exitCode == 0L then
+            if archName == "x86_64" || archName == "aarch64" then
+              "x64"
+            else if archName == "arm64" then
+              "arm64"
+            else
+              Builtin.printLine $"Unsupported architecture {archName}"
+              "Unsupported"
+
+          else
+            Builtin.printLine "Failed to get architecture"
+            arch.exitCode
```

Comment:
I think it's worth tidying this up a bit to represent these unexpected things as Results -- it's easy to lose context of the fact that these strings might be some negative case. I think there are a few places here where this is relevant